[NEMO-398] ExecutorRepresenter interface and DefaultExecutorRepresenter

alapha23 opened the pull request:
JIRA: [NEMO-398: ExecutorRepresenter interface and LambdaExecutorRepresenter ](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-398)

**Major Changes:**
`DefaultExecutorRepresenter` has the same function as previous `ExecutorRepresenter` does.
Now `DefaultExecutorRepresenter` implements `ExecutorRepresenter` interface.

**Tests for the changes:**
- Not any

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/ContainerManager.java:
```diff
@@ -192,13 +192,14 @@ public void onContainerAllocated(final String executorId,
     // Create the executor representation.
-    final ExecutorRepresenter executorRepresenter =
-      new ExecutorRepresenter(executorId, resourceSpec, messageSender, activeContext, serializationExecutorService,
+    final DefaultExecutorRepresenter defaultExecutorRepresenter =
```
final ExecutorRepresenter

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/ContainerManager.java:
```diff
@@ -170,7 +170,7 @@ public void onContainerAllocated(final String executorId,
    * @return a representation of the executor. (return an empty Optional if terminated)
    */
-  public Optional<ExecutorRepresenter> onContainerLaunched(final ActiveContext activeContext) {
+  public Optional<DefaultExecutorRepresenter> onContainerLaunched(final ActiveContext activeContext) {
```
ExecutorRepresenter

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/RuntimeMaster.java:
```diff
@@ -355,7 +357,7 @@ public void onContainerAllocated(final String executorId,
   public boolean onExecutorLaunched(final ActiveContext activeContext) {
     final Callable<Boolean> processExecutorLaunchedEvent = () -> {
-      final Optional<ExecutorRepresenter> executor = containerManager.onContainerLaunched(activeContext);
+      final Optional<DefaultExecutorRepresenter> executor = containerManager.onContainerLaunched(activeContext);
```
ExecutorRepresenter

taegeonum commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/MetricFlushTest.java:
```diff
@@ -119,8 +119,8 @@ public void onMessageWithContext(Object message, MessageContext messageContext)
-  private static ExecutorRepresenter newWorker(final MessageSender masterToWorkerSender) {
-    final ExecutorRepresenter workerRepresenter = mock(ExecutorRepresenter.class);
+  private static DefaultExecutorRepresenter newWorker(final MessageSender masterToWorkerSender) {
+    final DefaultExecutorRepresenter workerRepresenter = mock(DefaultExecutorRepresenter.class);
```
final ExecutorRepresenter

taegeonum commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/MetricFlushTest.java:
```diff
@@ -119,8 +119,8 @@ public void onMessageWithContext(Object message, MessageContext messageContext)
 
-  private static ExecutorRepresenter newWorker(final MessageSender masterToWorkerSender) {
-    final ExecutorRepresenter workerRepresenter = mock(ExecutorRepresenter.class);
+  private static DefaultExecutorRepresenter newWorker(final MessageSender masterToWorkerSender) {
```
ExecutorRepresenter

taegeonum commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/MetricFlushTest.java:
```diff
@@ -72,16 +72,16 @@ public void test() throws InjectionException, ExecutionException, InterruptedExc
     doAnswer((Answer<Void>) invocationOnMock -> {
-      final Consumer<Set<ExecutorRepresenter>> consumer = (Consumer) invocationOnMock.getArguments()[0];
-      consumer.accept(executorRepresenterSet);
+      final Consumer<Set<DefaultExecutorRepresenter>> consumer = (Consumer) invocationOnMock.getArguments()[0];
```
ExecutorRepresenter

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/ExecutorRepresenter.java:
```diff
@@ -18,248 +18,90 @@
-    this.serializationExecutorService = serializationExecutorService;
-    this.nodeName = nodeName;
-  }
+@DefaultImplementation(DefaultExecutorRepresenter.class)
```
please remove this. As DefaultExecutorRepresentor does not use @Inject, you cannot use this.

alapha23 replied:
@Inject added

alapha23 replied:
removed

alapha23 commented:
@taegeonum 
Updated. Removed most use of `DefaultExecutorRepresenter` in tests, and redundant log messages.
Note that instantiating an object would still need `DefaultExecutorRepresenter`, for instance [this line](https://github.com/apache/incubator-nemo/pull/227/files#diff-ce86ff9b39e9ec63431f873f5ef46670R105)

[JIRA](https://issues.apache.org/jira/browse/NEMO-398) added comments with the design logic with this pr.

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/ExecutorRepresenter.java:
```diff
@@ -18,248 +18,88 @@
- * c) Tasks scheduled/launched for the executor.
- * d) Name of the physical node which hosts this executor.
- * e) (Please add other information as we implement more features).
+ * Interface for DefaultExecutorRepresenter and LambdaExecutorRepresenter.
```
This is not a description of `ExecutorRepresenter`. Could you please add some comments what the `ExecutorRepresenter` is?

alapha23 replied:
Then I will keep the comments we had previously.

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/ExecutorRepresenter.java:
```diff
@@ -18,248 +18,95 @@
- * <p>
- * Contains information/state regarding an executor.
+ * Interface for DefaultExecutorRepresenter and LambdaExecutorRepresenter.
+ * Contains information/state regarding ann executor
```
typo

alapha23 replied:
Fixed.

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/DefaultExecutorRepresenter.java:
```diff
@@ -0,0 +1,278 @@
+import java.util.stream.Stream;
+
+/**
+ * (WARNING) This class is not thread-safe, and thus should only be accessed through ExecutorRegistry.
```
please add implementation specific info here, this comment is the same as the interface comment

alapha23 replied:
Done. I kept it brief. Anything else we have to say here?

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/ExecutorRepresenter.java:
```diff
@@ -18,248 +18,95 @@
- * (WARNING) This class is not thread-safe, and thus should only be accessed through ExecutorRegistry.
- * <p>
- * Contains information/state regarding an executor.
+ * Interface for DefaultExecutorRepresenter and LambdaExecutorRepresenter.
```
This interface doesn't have to be for the two specific implementation.  
Please remove this comment.

alapha23 replied:
True. Removed.

alapha23 commented:
@taegeonum @johnyangk 
Requested changes applied.
Mind taking a look again?

taegeonum commented on runtime/master/src/main/java/org/apache/nemo/runtime/master/resource/DefaultExecutorRepresenter.java:
```diff
@@ -0,0 +1,272 @@
+/**
+ * (WARNING) This class is not thread-safe, and thus should only be accessed through ExecutorRegistry.
+ * <p>
+ * Implements ExecutorRepresenter along with LambdaExecutorRepresenter.
```
Implements ExecutorRepresenter that communicates with Executors running on traditional resources (e.g., virtual machines or cluster resources)

alapha23 replied:
done
